### PR TITLE
Feature integration

### PR DESCRIPTION
**Pull request by Eagle1456**
Add integration functionality for the following rules:
- Constant rule
- Single variable same rule
- Multiplication by constant rule
- Power rule
- Sum rule
- Difference rule

### CODE REVIEW

File: src/Multiply.cpp [hunk 1]
```diff
@@ -338,4 +338,25 @@ auto Multiply<Expression>::Specialize(const Expression& other, tf::Subflow& subf
...
+{
+    // Single integration variable
+    if (auto variable = Variable::Specialize(integrationVariable); variable != nullptr) {
+        auto simplifiedMult = this->Simplify();
+
+        // Constant case - Constant number multiplied by integrand
+        if (auto constant = Multiply<Real, Expression>::Specialize(*simplifiedMult); constant != nullptr) {
+            auto exp = constant->GetLeastSigOp().Copy();
+            auto num = constant->GetMostSigOp();
+            auto integrated = (*exp).Integrate(integrationVariable);
```
Comment (matthew-mccall) [inline src/Multiply.cpp:351]: Why not use the arrow syntax?

**Review by matthew-mccall — CHANGES_REQUESTED** [review]
Looks good for a start. A couple things to change:
1. Use the declarative syntax and then call `Simplify` on the temporary. This should make the code more readable than use `std::make_unique`
2. In some cases you return a `Copy()` of the Expression when mathematically speaking it should be in the integrand of an integral. So we need to add an `Integral` binary node.
3. You have a lot of code nested in conditionals that check if the integrating variable is a variable. Consider using early returns instead of nesting code.

I understand these may not have been clarified before, so I will be sure to add them the guidelines now.

**Comment by Eagle1456** [comment]
@matthew-mccall, to address the changes requested:

1. Resolved
2. Resolved with assistance of @LevyLCH 
3. The only thing is that some of the classes use the Variable.GetName() function which can only be discerned if the Variable class is specialized.

File: include/Oasis/Integrate.hpp
Comment (matthew-mccall) [inline include/Oasis/Integrate.hpp:1]: Rename the class to `Integral`

File: include/Oasis/Expression.hpp
Comment (matthew-mccall) [inline include/Oasis/Expression.hpp:1]: Rename the function to `Integrate` to be consistent

File: src/Integrate.cpp
Comment (matthew-mccall) [inline src/Integrate.cpp:1]: The `Integration` logic should be inside the `Add`, `Exponent`, etc. classes. The `Integral` class is meant to be a wrapper such that calling `Simplify` on an integral is what performs the integration by calling `Integrate` on the operands.

**Review by matthew-mccall — CHANGES_REQUESTED** [review]

File: include/Oasis/Expression.hpp [hunk 2]
```diff
@@ -181,6 +182,19 @@ class Expression {
...
+     * @return An indefinite integral of the expression added to a constant
+     */
+    [[nodiscard]] virtual auto Integrate(const Expression&) -> std::unique_ptr<Expression>;
+
+    /**
+     * Attempts to integrate this expression using integration rules
+     *
+     * @return A solved definite integral of the expression
+     */
+    [[nodiscard]] virtual auto Integrate(const Expression&, const Expression&, const Expression&) -> std::unique_ptr<Expression>;
```
Comment (matthew-mccall) [inline include/Oasis/Expression.hpp:197]: Rename this to IntegrateWithBounds as to portray clearer intent

**Review by matthew-mccall — CHANGES_REQUESTED** [review]

**Comment by matthew-mccall** [comment]
@Eagle1456  Hi, thank you for picking up on the changes with regards to refactoring out the `ToString` and `MathML` functions. I will be required `Oasis::Extras` to be able to build successfully soon, so please look out for when that gets merged in, and make sure to implement the necessary serialization functions. (I see you declared them, but not defined them). Also rename the definite `Integrate` function to `IntegrateWithBounds`. Other than that, looks good to me. I will merge once those changes are made.

**Review by matthew-mccall — APPROVED** [review]

**Pull request merged**